crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -1076,6 +1076,133 @@ impl Mesh {
+    ///
+    /// Returns an error if any of the following conditions are met (see [`MeshTrianglesError`]):
+    /// * The Mesh's [primitive topology] is not `TriangleList` or `TriangleStrip`.
+    /// * The Mesh is missing position or index data.
+    /// * The Mesh's position data has the wrong format (not `Float32x3`).
+    ///
+    /// [primitive topology]: PrimitiveTopology
+    pub fn triangles(&self) -> Result<impl Iterator<Item = Triangle3d> + '_, MeshTrianglesError> {
+        let Some(position_data) = self.attribute(Mesh::ATTRIBUTE_POSITION) else {
+            return Err(MeshTrianglesError::MissingPositions);
+        };
+
+        let Some(vertices) = position_data.as_float3() else {
+            return Err(MeshTrianglesError::PositionsFormat);
+        };
+
+        let Some(indices) = self.indices() else {
+            return Err(MeshTrianglesError::MissingIndices);
+        };
+
+        match self.primitive_topology {
+            PrimitiveTopology::TriangleList => {
+                // When indices reference out-of-bounds vertex data, the triangle is omitted.
+                // This implicitly truncates the indices to a multiple of 3.
+                let iterator = match indices {
+                    Indices::U16(vec) => FourIterators::First(
+                        vec.as_slice()
+                            .chunks_exact(3)
+                            .flat_map(move |indices| indices_to_triangle(vertices, indices)),
+                    ),
+                    Indices::U32(vec) => FourIterators::Second(
+                        vec.as_slice()
+                            .chunks_exact(3)
+                            .flat_map(move |indices| indices_to_triangle(vertices, indices)),
+                    ),
+                };
+
+                return Ok(iterator);
+            }
+
+            PrimitiveTopology::TriangleStrip => {
+                // When indices reference out-of-bounds vertex data, the triangle is omitted.
+                // If there aren't enough indices to make a triangle, then an empty vector will be
+                // returned.
+                let iterator = match indices {
+                    Indices::U16(vec) => FourIterators::Third(
+                        vec.as_slice()
+                            .windows(3)
+                            .flat_map(move |indices| indices_to_triangle(vertices, indices)),
+                    ),
+                    Indices::U32(vec) => FourIterators::Fourth(
+                        vec.as_slice()
+                            .windows(3)
+                            .flat_map(move |indices| indices_to_triangle(vertices, indices)),
+                    ),
+                };
+
+                return Ok(iterator);
+            }
+
+            _ => {
+                return Err(MeshTrianglesError::WrongTopology);
+            }
+        };
+
+        fn indices_to_triangle<T: TryInto<usize> + Copy>(
```

Comment:
Given that this method is in a hot-loop of the iterator, would it make sense to mark #[inline]?

Reply:
Maybe. This is super hard to predict in my experience.